src/AutoRest.CSharp.V3/Output/Builders/SerializationBuilder.cs

```diff
@@ -163,33 +168,54 @@ private XmlObjectSerialization BuildXmlObjectSerialization(ObjectSchema objectSc
                 );
         }
 
-        private IEnumerable<JsonPropertySerialization> GetPropertySerializationsFromBag(PropertyBag propertyBag, ObjectType objectType)
+        private IEnumerable<JsonPropertySerialization> GetPropertySerializationsFromBag(PropertyBag propertyBag, ObjectType objectType, VirtualParameter[] flattenedParameters)
         {
             foreach (Property property in propertyBag.Properties)
             {
-                string propertyName = objectType.GetPropertyForSchemaProperty(property, includeParents: true).Declaration.Name;
+                CodeWriterDelegate? propertyName = null;
+                if (flattenedParameters.Any())
+                {
+                    VirtualParameter? virtualParameter = flattenedParameters.FirstOrDefault(fp => fp.TargetProperty == property);
+                    if (virtualParameter != null)
+                    {
+                        propertyName = virtualParameter.Schema is ConstantSchema constantSchema
+                            ? w => w.WriteConstant(BuilderHelpers.ParseConstant(constantSchema, _typeFactory))
+                            : (CodeWriterDelegate)(w => w.Append($"{virtualParameter.Language.Default.Name}"));
+                    }
+                }
```

Comment:
Now, since we're going through the properties, I simply find the virtual parameter that matches the property, and extract the correct name from it.